File: backend/src/main/java/com/demo/travel_expense_management/legalization/mapper/LegalizationMapper.java
package com.demo.travel_expense_management.legalization.mapper;

import java.util.List;

import org.mapstruct.Mapper;
import org.mapstruct.Mapping;

import com.demo.travel_expense_management.expense.mapper.ExpenseMapper;
import com.demo.travel_expense_management.legalization.dto.response.LegalizationResponse;
import com.demo.travel_expense_management.legalization.dto.response.LegalizationSummaryResponse;
import com.demo.travel_expense_management.legalization.entity.Legalization;

@Mapper(componentModel = "spring", uses = {ExpenseMapper.class})
public interface LegalizationMapper {

    @Mapping(target = "travelRequestId", source = "travelRequest.id")
    @Mapping(target = "requestNumber", source = "travelRequest.requestNumber")
    @Mapping(target = "applicantName", expression = "java(legalization.getTravelRequest().getApplicant().getFirstName() + \" \" + legalization.getTravelRequest().getApplicant().getLastName())")
    @Mapping(target = "costCenterId", source = "costCenter.id")
    @Mapping(target = "costCenterName", source = "costCenter.name")
    @Mapping(target = "totalExpenses", expression = "java(legalization.getExpenses() == null ? 0 : legalization.getExpenses().stream().mapToInt(e -> e.getAmount()).sum())")
    @Mapping(target = "status", source = "travelRequest.status")
    LegalizationResponse toResponse(Legalization legalization);

    @Mapping(target = "requestNumber", source = "travelRequest.requestNumber")
    @Mapping(target = "applicantName", expression = "java(legalization.getTravelRequest().getApplicant().getFirstName() + \" \" + legalization.getTravelRequest().getApplicant().getLastName())")
    @Mapping(target = "totalExpenses", expression = "java(legalization.getExpenses() == null ? 0 : legalization.getExpenses().stream().mapToInt(e -> e.getAmount()).sum())")
    @Mapping(target = "status", source = "travelRequest.status")
    LegalizationSummaryResponse toSummaryResponse(Legalization legalization);

    List<LegalizationSummaryResponse> toSummaryResponseList(List<Legalization> legalizations);
}

File: frontend/src/lib/mapper/legalization.mapper.ts
import type {
	ExpenseResponseDto,
	LegalizationResponseDto,
} from '$lib/dto/legalization';
import type { Expense, Legalization } from '$lib/models/legalization';

export function toExpense(dto: ExpenseResponseDto): Expense {
	return {
		id: dto.id,
		expenseTypeId: dto.expenseTypeId,
		expenseTypeName: dto.expenseTypeName,
		expenseDate: dto.expenseDate,
		description: dto.description,
		amount: dto.amount,
	};
}

export function toLegalization(dto: LegalizationResponseDto): Legalization {
	return {
		id: dto.id,
		travelRequestId: dto.travelRequestId,
		requestNumber: dto.requestNumber,
		applicantName: dto.applicantName,
		costCenterId: dto.costCenterId,
		costCenterName: dto.costCenterName,
		totalExpenses: dto.totalExpenses,
		status: dto.status,
		submittedAt: dto.submittedAt,
		expenses: dto.expenses.map(toExpense),
	};
}
